File: geopdes_base/inst/operators/op_f_v.m
% OP_F_V: assemble the right-hand side vector r = [r(i)], with  r(i) = (f, v_i).
%
%   rhs = op_f_v (spv, msh, coeff);
%
% INPUT:
%     
%   spv:   structure representing the function space (see sp_bspline_2d_phys)
%   msh:   structure containing the domain partition and the quadrature rule (see msh_push_forward_2d)
%   coeff: source function evaluated at the quadrature points
%
% OUTPUT:
%
%   rhs: assembled right-hand side
% 
% Copyright (C) 2009, 2010 Carlo de Falco
% Copyright (C) 2011 Rafael Vazquez
%
%    This program is free software: you can redistribute it and/or modify
%    it under the terms of the GNU General Public License as published by
%    the Free Software Foundation, either version 3 of the License, or
%    (at your option) any later version.

%    This program is distributed in the hope that it will be useful,
%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%    GNU General Public License for more details.
%
%    You should have received a copy of the GNU General Public License
%    along with this program.  If not, see <http://www.gnu.org/licenses/>.



function rhs = op_f_v2 (spv, msh, coeff, element_list)
  
 coeff = reshape (coeff, spv.ncomp, msh.nqn, msh.nel);

 if (nargin == 3)
   element_list = 1:msh.nel;
 end
 nel = numel (element_list);

 rhs   = zeros (spv.ndof, 1);
 shpv  = reshape (spv.shape_functions, spv.ncomp, msh.nqn, spv.nsh_max, nel);

 for iel = 1:nel
   iel_glob = element_list (iel);
   if (all (msh.jacdet(:,iel_glob)))
     rhs_loc = zeros (spv.nsh(iel), 1);

     jacdet_weights = msh.jacdet(:, iel_glob) .* msh.quad_weights(:, iel_glob);
     jacdet_weights = repmat (jacdet_weights, [1, spv.nsh(iel)]);

     coeff_iel = repmat (coeff(:, :, iel_glob), [1, 1, spv.nsh(iel)]);

     shpv_iel = reshape (shpv(:, :, 1:spv.nsh(iel), iel), spv.ncomp, msh.nqn, spv.nsh(iel));

     val1 = reshape (sum (shpv_iel .* coeff_iel, 1), [msh.nqn, spv.nsh(iel)]);
     rhs_loc = sum (jacdet_weights .* val1, 1).';

     rhs(spv.connectivity(1:spv.nsh(iel), iel)) = rhs(spv.connectivity(1:spv.nsh(iel), iel)) + rhs_loc; 
   else
     warning ('geopdes:jacdet_zero_at_quad_node', 'op_f_v: singular map in element number %d', iel)
   end
 end
 
end

File: trunk/geopdes_base/src/op_f_v.cc
/* Copyright (C) 2009 Carlo de Falco
   Copyright (C) 2011 Rafael Vazquez

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.
 
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
*/


#include "geopdes.h"

DEFUN_DLD(op_f_v, args, nargout,"\n\
OP_F_V: assemble the right-hand side vector r = [r(i)], with  r(i) = (f, v_i). \n\
\n\
  mat = op_f_v (spv, msh, coeff); \n\
\n\
INPUT:\n\
\n\
  spv:   structure representing the function space (see sp_bspline_2d_phys) \n\
  msh:   structure containing the domain partition and the quadrature rule (see msh_push_forward_2d) \n\
  coeff: source function evaluated at the quadrature points \n\
\n\
OUTPUT:\n\
\n\
  mat: assembled right-hand side \n\
")
{
  
  octave_value_list retval;

  geopdes_mesh  msh (args(1).map_value ());
  geopdes_space sp  (args(0).map_value (), msh);

  if (!error_state)
    {

      const octave_idx_type nel = msh.nel (), ncomp = sp.ncomp (), nqn = msh.nqn ();

      dim_vector    idx (ncomp, nqn, nel);
      NDArray       coeff = args(2).array_value ().reshape (idx);
      ColumnVector mat (sp.ndof (), 0.0);

      octave_idx_type counter = 0, iel, inode, idof, icmp;

#pragma omp parallel default (none) shared (msh, sp, idx, mat, coeff)
      {
        double local_contribution;

#pragma omp for
      for ( iel=0; iel < nel; iel++) 
        if (msh.area (iel) > 0.0)
	  {
            const octave_idx_type nsh = sp.nsh (iel);
            double jacdet_weights[nqn];

            for ( inode = 0; inode < nqn; inode++)
              {
                jacdet_weights[inode] = msh.jacdet (inode, iel) *
                  msh.weights (inode, iel);
              }

            double shp[nsh][nqn][ncomp];
            int conn[nsh];

            for ( idof = 0; idof < nsh; idof++) 
              {
                for ( inode = 0; inode < nqn; inode++)
                  {
                    for ( icmp = 0; icmp < ncomp; icmp++)
                      {
                        shp[idof][inode][icmp] = sp.shape_functions (icmp, inode, idof, iel);
                      }
                  }
                conn[idof] = sp.connectivity (idof, iel) - 1;
              }


            for ( idof = 0; idof < nsh; idof++) 
	      {
                for ( inode = 0; inode < nqn; inode++)
                  {
                    if (msh.weights (inode, iel) > 0.0)
                      {
                        double s = 0.0;
                        for ( icmp = 0; icmp < ncomp; icmp++)
                          s += shp[idof][inode][icmp] * coeff (icmp, inode, iel);
                        local_contribution = jacdet_weights[inode] * s;
#pragma omp critical
                        mat(conn[idof]) += local_contribution;			  
                      }  
                  } // end for inode
	      } // end for idof
          } else {
#pragma omp critical
          {warning_with_id ("geopdes:zero_measure_element", "op_f_v: element %d has 0 area (or volume)", iel);}
        }  // end for iel, if area > 0
      }  // end for parallel region
      retval(0) = octave_value (mat);
    } // end if !error_state
  return retval;
}

